common/utils/budget.js

```diff
@@ -975,18 +1008,89 @@ export const sumShareCostsForFundingSource = ({
       updatedBudget[fundingSource].combined.total.medicaid += totalMedicaidCost;
     }
     if (fundingSource !== 'hitAndHie') {
-      updatedBudget.combined[year].federal += totalMedicaidCostShares.fedShare;
-      updatedBudget.combined.total.federal += totalMedicaidCostShares.fedShare;
-      updatedBudget.combined[year].state += totalMedicaidCostShares.stateShare;
-      updatedBudget.combined.total.state += totalMedicaidCostShares.stateShare;
-      updatedBudget.combined[year].medicaid += totalMedicaidCost;
-      updatedBudget.combined.total.medicaid += totalMedicaidCost;
+      // Add to the budget's overall combined totals
+      updatedBudget = sumShareCostsForCombinedTotals({
+        budget: updatedBudget,
+        year,
+        medicaidShare: totalMedicaidCost,
+        fedShare: totalMedicaidCostShares.fedShare,
+        stateShare: totalMedicaidCostShares.stateShare
+      });
     }
   }
 
   return updatedBudget;
 };
 
+/**
+ * Sum up the federal, state, and medicaid costs for the activity:
+ * - by funding category (ddi, mando) and cost category ('contractors', 'expenses', 'statePersonnel')
+ *     for each year and the total of all the years
+ * - by funding category for each year and the total of all the years
+ * @param {Object} budget The budget object
+ * @param {Number} year The FFY
+ * @param {String} fundingCategory The funding category key name (i.e. 'ddi', 'mando')
+ * @param {Object} activityTotals The activity total object
+ * @param {Object} costCategoryShare The federal, state, and medicaid share of the total costs
+ * @returns {Object} the updated budget
+ */
+export const sumShareCostsForFundingCategory = ({
+  budget,
+  year,
+  fundingCategory,
+  activityTotals,
+  costCategoryShare
+} = {}) => {
+  const updatedBudget = deepCopy(budget);
+  if (
+    budget &&
+    year &&
+    fundingCategory &&
+    activityTotals &&
+    costCategoryShare
+  ) {
+    ['contractors', 'expenses', 'statePersonnel'].forEach(costCategory => {
+      if (
+        updatedBudget?.[fundingCategory]?.[costCategory]?.[year] !== undefined
+      ) {
+        const costTotal = activityTotals.data[costCategory][year];
+        const medicaidShare = costCategoryShare.medicaidShare[costCategory];
+        const fedShare = costCategoryShare.fedShare[costCategory];
+        const stateShare = costCategoryShare.stateShare[costCategory];
+
+        // add to costCategory for the particular year (for particular funding category)
+        updatedBudget[fundingCategory][costCategory][year].total += costTotal;
```

Comment:
this is where we also need to store it under fed-state split, it should be
```
updatedBudget[fundingCategory][fedSplit][costCategory][year].total += costTotal;
```
all the way down, whenever there is [fundingCategory] it should be followed by [fedSplit]